File: 1081. Smallest Subsequence of Distinct Characters.py
# https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/

# Return the lexicographically smallest subsequence of s that contains all the
# distinct characters of s exactly once.

# Note: This question is the same as 316: https://leetcode.com/problems/remove-
# duplicate-letters/

################################################################################

# record last postion of each char
# use stack and pop previous chars when i) new char is smaller and ii) we can add the popped char back later

class Solution:
    def smallestSubsequence(self, s: str) -> str:
        last_pos = {}
        for idx, char in enumerate(s):
            last_pos[char] = idx
        
        stack = []
        for idx, char in enumerate(s):
            if char not in stack:
                # pop the previous chars if the new char is smaller
                # but only when we can add the popped char back: idx < last_pos[popped_char]
                while stack and char < stack[-1] and idx < last_pos[stack[-1]]:
                    stack.pop()
                stack.append(char)
        return ''.join(stack)

File: 1081. Smallest Subsequence of Distinct Characters.cpp
// https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/

// Return the lexicographically smallest subsequence of s that contains all the
// distinct characters of s exactly once.

// Note: This question is the same as 316: https://leetcode.com/problems/remove-
// duplicate-letters/

////////////////////////////////////////////////////////////////////////////////

class Solution {
public:
    string smallestSubsequence(string s) {
        int n = s.size();
        unordered_map<char, int> lastPos;
        unordered_map<char, bool> added;
        for (int i = 0; i < n; ++i) lastPos[s[i]] = i;
        
        string ans;
        for (int i = 0; i < n; ++i) {
            if (!added[s[i]]) {
                while (!ans.empty() and ans.back() > s[i] and lastPos[ans.back()] > i) {
                    added[ans.back()] = false; ans.pop_back();
                }
                ans += s[i]; added[s[i]] = true;
            }
        }
        return ans;
    }
};
